MUTATION-SCOPE.md

```diff
@@ -0,0 +1,439 @@
+AsyncContext value, it needs non-trivial refactor:
+
+```js
+const asyncVar = new AsyncContext.Context();
+
+function *gen() {
+  const span = createSpan();
+  yield asyncVar.run(span, () => computeResult());
+  yield asyncVar.run(span, () => computeResult2());
+  // ...or
+  yield* asyncVar.run(span, function *() {
+    yield computeResult();
+    yield computeResult2();
+  });
+}
+```
+
+`.run(val, fn)` creates a new function body. The new function environment
+is not equivalent to the outer environment and can not trivially share code
+fragments between them. Additionally, `break`/`continue`/`return` can not be
+refactored naively.
+
+It will be more intuitive to be able to insert a new line and without refactor
+existing code snippet.
+
+```diff
+ const asyncVar = new AsyncContext.Variable();
+
+ function *gen() {
++  using _ = asyncVar.withValue(createSpan(i));
+   yield computeResult(i);
+   yield computeResult2(i);
+ }
+```
+
+## The set semantic with scope enforcement
+
+With the name of `set`, this method actually doesn't modify existing async
+context snapshots, similar to consecutive `run` operations. For example, in
+the following case, `set` doesn't change the context variables in async tasks
+created just prior to the mutation:
+
+An alternative to exposing the `set` semantics directly is allowing mutation
+with well-known symbol interface [`@@dispose`][] by using declaration (and
+potentially enforcing the `using` declaration with [`@@enter`][]).
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+{
+  using _ = asyncVar.withValue("main");
+  new AsyncContext.Snapshot() // snapshot 0
+  console.log(asyncVar.get()); // => "main"
+}
+
+{
+  using _ = asyncVar.withValue("value-1");
+  new AsyncContext.Snapshot() // snapshot 1
+  Promise.resolve()
+    .then(() => { // continuation 1
+      console.log(asyncVar.get()); // => 'value-1'
+    })
+}
+
+{
+  using _ = asyncVar.withValue("value-2");
+  new AsyncContext.Snapshot() // snapshot 2
+  Promise.resolve()
+    .then(() => { // continuation 2
+      console.log(asyncVar.get()); // => 'value-2'
+    })
+}
+```
+
+The value mapping is equivalent to:
+
+```
+⌌-----------⌍ snapshot 0
+|   'main'  |
+⌎-----------⌏
+      |
+⌌-----------⌍ snapshot 1
+| 'value-1' |  <---- the continuation 1
+⌎-----------⌏
+      |
+⌌-----------⌍ snapshot 2
+| 'value-2' |  <---- the continuation 2
+⌎-----------⌏
+```
+
+Each `@@enter` operation create a new value slot preventing any mutation to
+existing snapshots where the current `AsyncContext.Variable`'s value was
+captured.
+
+This trait is important with both `run` and `set` because mutations to
+`AsyncContext.Variable`s must not mutate prior `AsyncContext.Snapshot`s.
+
+> Note: this also applies to [`ContinuationVariable`][]
+
+However, the well-known symbol `@@dispose` and `@@enter` is not bound to the
+`using` declaration syntax, and they can be invoked manually. This can be a
```

Comment:
Allowing a `@@enter` without `@@dispose` opens the door for leaks. Either by forgetting `@@dispose` or calling them in the wrong order.
In my opinion the major benefit of a scoped `set` is gone because of this.

Reply:
`@@enter` and `@@dispose` are expected to be used with `using` declaration. Allowing it to be invoked manually is primarily for library extension, and shown in example below that compositing active spans with the context variables.

Reply:
As long as they are used correct in objects which are again intended for use with `using` and as a result the correct nesting is kept it should be fine.
If it is easy to mix up sequence of `@@dispose` it's quite problematic.